### PR TITLE
performance metrics collector

### PR DESCRIPTION
**Pull request by wangmingliang-ms**

### CODE REVIEW

File: Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/task/AzureTaskContext.java [hunk 1]
```diff
@@ -75,33 +82,50 @@ private AzureTaskContext(final long threadId, final Deque<AzureOperationRef> ope
...
+        @Setter(AccessLevel.PACKAGE)
+        private AzureTask<?> task;
 
         private Node(final AzureTaskContext parent) {
             super();
             this.parent = parent;
         }
 
-        void pushOperation(final AzureOperationRef operation) {
+        public AzureTaskContext getRealParent() {
```
Comment (andxu) [inline Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/task/AzureTaskContext.java:115]: getOriginalParent?

File: Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+import lombok.SneakyThrows;
+
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+public class AzurePerformanceMetricsCollector {
+
+    public static final String NONE = "NONE";
```
Comment (andxu) [inline Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java:41]: private?

File: Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+            final AzureOperation annotation = AzureOperationUtils.getAnnotation(current);
+            if (annotation.type() == AzureOperation.Type.ACTION) {
+                break; // should not end side-effect action.
+            }
+        }
+    }
+
+    private static void enterOperation(final AzureOperationRef current, final Deque<? extends AzureOperationRef> ancestors) {
+        final long timestamp = System.currentTimeMillis();
+        final HashMap<String, String> properties = buildOperationProperties(current, ancestors);
```
Comment (andxu) [inline Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java:74]: HashMap<String, String> ->  Map<String, String>

File: Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+    public static void beforeEnter(final AzureTask<?> task) {
+        AzurePerformanceMetricsCollector.enterTask(task);
+    }
+
+    public static void afterExit(final AzureTask<?> task) {
+        AzurePerformanceMetricsCollector.exitTask(task);
+        final AzureTaskContext.Node context = task.getContext();
+        final Deque<AzureOperationRef> operations = AzureTaskContext.getContextOperations(context);
+        while (!operations.isEmpty()) { // end operations asynchronously
```
Comment (andxu) [inline Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java:62]: why we need to end a lot of operations? end current operation is enough?

File: Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        final HashMap<String, String> properties = buildOperationProperties(current, ancestors);
+        properties.put("timestamp", String.valueOf(timestamp));
+        properties.put("action", "END");
+        sendTelemetry(properties);
+    }
+
+    private static void enterTask(final AzureTask<?> task) {
+        final long timestamp = System.currentTimeMillis();
+        final HashMap<String, String> properties = buildTaskProperties(task);
+        properties.put("timestamp", String.valueOf(timestamp));
```
Comment (andxu) [inline Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java:99]: timestamp to constant

File: Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+        properties.put("action", "EXIT");
+        sendTelemetry(properties);
+    }
+
+    private static void endOperation(final AzureOperationRef current, final Deque<? extends AzureOperationRef> ancestors) {
+        final long timestamp = System.currentTimeMillis();
+        final HashMap<String, String> properties = buildOperationProperties(current, ancestors);
+        properties.put("timestamp", String.valueOf(timestamp));
+        properties.put("action", "END");
```
Comment (andxu) [inline Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java:92]: action values: enter end? give a better name to replace "action"

File: Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/task/AzureTaskContext.java (hunk 1 above)
Comment (wangmingliang-ms) [reply Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/task/AzureTaskContext.java]: removed already

File: Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java (hunk 2 above)
Comment (wangmingliang-ms) [reply Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java]: fixed

File: Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java (hunk 3 above)
Comment (wangmingliang-ms) [reply Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java]: fixed

File: Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java (hunk 4 above)
Comment (wangmingliang-ms) [reply Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java]: this is to actually make it fast/fast in later summarizing, it's no longer needed with the new id schema.

File: Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java (hunk 5 above)
Comment (wangmingliang-ms) [reply Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java]: fixed

File: Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java (hunk 6 above)
Comment (wangmingliang-ms) [reply Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java]: remove, we now only have `ENTER`/`EXIT`

File: Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class AzurePerformanceMetricsCollector {
+    private static final File file = new File(System.getProperty("user.home") + "/performance.csv");
```
Comment (andxu) [inline Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java:47]: new File(System.getProperty("user.home"), "performance.csv")

Comment (andxu) [reply Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java:47]: I suggest to add a folder named '.azuretools'

File: Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+    private static final String TELEMETRY_OP_PARENT_ID = "parent_id";
+    private static final String TELEMETRY_OP_NAME = "name";
+    private static final String TELEMETRY_OP_TYPE = "type";
+    private static final String OP_ACTION_CREATE = "CREATE";
+    private static final String OP_ACTION_ENTER = "ENTER";
+    private static final String OP_ACTION_EXIT = "EXIT";
+
+    public static void afterCreate(final IAzureOperation op) {
+        final long timestamp = System.currentTimeMillis();
+        final Map<String, String> properties = buildProperties(op);
```
Comment (andxu) [inline Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java:63]: duplicate exception:        properties.put(TELEMETRY_OP_ACTION, OP_ACTION_CREATE);
please wrap it in a private common method

File: Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+        final String val = String.format("%s, %s, %s, %s, %s, %s, %s",
+            properties.get(TELEMETRY_OP_CONTEXT_ID),
+            properties.get(TELEMETRY_OP_TIMESTAMP),
+            properties.get(TELEMETRY_OP_ID),
+            properties.get(TELEMETRY_OP_NAME),
+            properties.get(TELEMETRY_OP_TYPE),
+            properties.get(TELEMETRY_OP_ACTION),
+            properties.get(TELEMETRY_OP_PARENT_ID)
+        );
+        FileUtils.writeStringToFile(file, val + System.lineSeparator(), StandardCharsets.UTF_8, true);
```
Comment (andxu) [inline Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java:136]: shall we use https://www.baeldung.com/apache-commons-csv instead?

Comment (wangmingliang-ms) [reply Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java:136]: no, it's just for demo

File: Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java (hunk 7 above)
Comment (wangmingliang-ms) [reply Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java:47]: it's only for demo

File: Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java (hunk 8 above)
Comment (wangmingliang-ms) [reply Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java:63]: I don't understand.

**Review by andxu — APPROVED** [review]

**Pull request merged**